dashboard/app/helpers/proxy_helper.rb

```diff
@@ -66,6 +66,51 @@ def render_proxied_url(location, allowed_content_types:, allowed_hostname_suffix
     render_error_response 400, "Network error #{e.class} #{e.message}"
   end
 
+  # Unlike render_proxied_url, this does not attempt to render the URL, but instead
+  # just follows it to figure out what the redirected URL is. It's also the case
+  # that we'll stop as soon as we've been redirected to a URL on this site.
+  def resolve_redirect_url(location, allowed_hostname_suffixes:, redirect_limit: 5)
+    if redirect_limit == 0
+      return 500, 'Redirect loop'
+    end
+
+    url = URI.parse(location)
+    raise URI::InvalidURIError.new if url.host.nil? || url.port.nil?
+
+    # If we've resolved to a path on this host/port, stop trying to redirect
+    if url.host === request.host && url.port === request.port
+      return 200, location
+    end
+
+    # If hostname isn't in our white list, don't attempt a redirect
+    unless allowed_hostname?(url, allowed_hostname_suffixes)
+      return 200, location
+    end
+
+    http = Net::HTTP.new(url.host, url.port)
+    http.use_ssl = url.scheme == 'https'
+    path = (url.path.empty?) ? '/' : url.path
+    query = url.query || ''
+
+    # Limit how long we're willing to wait.
+    http.open_timeout = 3
+    http.read_timeout = 3
```

Comment:
Probably good to document units here.

3 seconds seems a little short.  Can the user easily retry the import if it fails the first time?